In [2]:
pip install feature-engine


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [4]:
pip install skrebate


  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for skrebate: filename=skrebate-0.62-py3-none-any.whl size=29355 sha256=acfb836448a47e4af6a6d47421ab1b52e1289aea8b3a782a75c82f35c3d161bb
  Stored in directory: c:\users\marti\appdata\local\pip\cache\wheels\8a\d2\67\8368fca718041057b33d6448ca526aaf1ce2e404c94b901e41
Successfully built skrebate
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [5]:
pip install skrub -U

   ---------------------------------------- 0.0/330.2 kB ? eta -:--:--
   ------------------------ --------------- 204.8/330.2 kB 4.1 MB/s eta 0:00:01
   ---------------------------------------  327.7/330.2 kB 4.1 MB/s eta 0:00:01
   ---------------------------------------- 330.2/330.2 kB 3.4 MB/s eta 0:00:00
   ---------------------------------------- 0.0/134.9 kB ? eta -:--:--
   ---------------------------------------  133.1/134.9 kB 2.6 MB/s eta 0:00:01
   ---------------------------------------- 134.9/134.9 kB 2.0 MB/s eta 0:00:00
   ---------------------------------------- 0.0/64.9 kB ? eta -:--:--
   ---------------------------------------- 64.9/64.9 kB 1.8 MB/s eta 0:00:00
   ---------------------------------------- 0.0/166.4 kB ? eta -:--:--
   ---------------------------------- ----- 143.4/166.4 kB 4.3 MB/s eta 0:00:01
   ---------------------------------------- 166.4/166.4 kB 2.5 MB/s eta 0:00:00
   ---------------------------------------- 0.0/102.4 kB ? eta -:--:--
   -----


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [7]:
pip install category_encoders


   ---------------------------------------- 0.0/85.7 kB ? eta -:--:--
   ---- ----------------------------------- 10.2/85.7 kB ? eta -:--:--
   --------- ------------------------------ 20.5/85.7 kB 640.0 kB/s eta 0:00:01
   -------------------------------------- - 81.9/85.7 kB 919.0 kB/s eta 0:00:01
   ---------------------------------------- 85.7/85.7 kB 802.9 kB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [9]:
pip install dirty_cat


Note: you may need to restart the kernel to use updated packages.
     ---------------------------------------- 0.0/103.0 kB ? eta -:--:--
     --- ------------------------------------ 10.2/103.0 kB ? eta -:--:--
     -------------- ---------------------- 41.0/103.0 kB 393.8 kB/s eta 0:00:01
     -----------------------------------  102.4/103.0 kB 845.5 kB/s eta 0:00:01
     ------------------------------------ 103.0/103.0 kB 742.2 kB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
   ---------------------------------------- 0.0/25.3 MB ? eta -:--:--
   ---------------------------------------- 0.2/25.3 MB 5.1 MB/s eta 0:00:05
   ---------------------------------------- 0.3/25.3 MB 3.5 MB/s eta 0


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [13]:
pip uninstall dirty_cat


^C
Note: you may need to restart the kernel to use updated packages.


In [1]:
pip install dirty_cat==0.1.9


Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement dirty_cat==0.1.9 (from versions: 0.0.1a0, 0.0.1b1, 0.0.1b2, 0.0.1b3, 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.7, 0.1.0, 0.2.0a1, 0.2.0, 0.2.1, 0.3.0b1, 0.3.0, 0.4.0b1, 0.4.0b2, 0.4.0, 0.4.1, 0.4.2, 0.4.2.post0)
ERROR: No matching distribution found for dirty_cat==0.1.9

[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [3]:
pip install dirty_cat


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:
pip show dirty_cat


Name: dirty_cat
Version: 0.4.2.post0
Summary: Machine learning with dirty categories.
Home-page: 
Author: Patricio Cerda
Author-email: patricio.cerda@inria.fr
License: BSD
Location: C:\Users\marti\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages
Requires: joblib, numpy, pandas, pyarrow, requests, scikit-learn, scipy
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [5]:
pip install pandas numpy scikit-learn skrub category_encoders


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:
pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [36]:
pip install category_encoders


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\marti\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [ ]:
import pandas as pd
import numpy as np
from xgboost import XGBRegressor
from feature_engine.selection import DropCorrelatedFeatures, DropDuplicateFeatures
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.model_selection import train_test_split

# Importar las clases 
from utils.transformations import MyTransformation
from utils.filters import MyFilter


In [2]:
# Cargar el dataset de entrenamiento
df_train = pd.read_csv("data/preprocessed/train_data.csv")
X_train, y_train = df_train.drop(columns=['Price']), df_train['Price']

In [3]:
# Crear e inicializar el preprocesador y el filtro de datos
preprocessor = MyTransformation(target_encode_cols=['city', 'Area'])  
filter = MyFilter(variance_threshold=0.01, corr_threshold=0.9, k_best=4)  

In [4]:
# Ajustar el preprocesador y el filtro a los datos de entrenamiento
preprocessor.fit(X_train, y_train)


X shape:  (20974, 40)
bin_vars_columns shape:  (36,)
low_card_columns shape:  37


In [5]:
# Transformar los datos de entrenamiento
X_processed, y_processed = preprocessor.transform(X_train, y_train)


X shape:  (20974, 40)
X_crossed_features shape:  (20974, 10)
X_EXPANDED shape:  (20974, 14)


In [6]:
filter.fit(X_processed, y_processed)


(20974, 14)
Shape después de baja varianza: (20974, 7)
Shape después de eliminar duplicados: (20974, 7)
Shape después de eliminar correlacionadas: (20974, 6)


In [7]:
X_filtered, y_filtered = filter.transform(X_processed, y_processed)

Shape después de baja varianza en la transformación: (20974, 7)
Shape después de eliminar duplicados en la transformación: (20974, 7)
Shape después de eliminar correlacionadas en la transformación: (20974, 6)


In [8]:
# Dividir los datos de entrenamiento en conjunto de entrenamiento y validación 
X_train_final, X_val, y_train_final, y_val = train_test_split(X_filtered, y_filtered, test_size=0.2, random_state=42)

In [9]:
# Entrenar el modelo XGBoost con los datos filtrados
xgboost_model = XGBRegressor(n_estimators=500, random_state=42)
xgboost_model.fit(X_train_final, y_train_final)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=500, n_jobs=None,
             num_parallel_tree=None, random_state=42, ...)

In [10]:
xgboost_model.score(X_filtered, y_filtered)

0.47388281931754583

In [11]:
#  Cargar el dataset de prueba
df_test = pd.read_csv("data/preprocessed/test_data.csv")
X_test, y_test = df_test.drop(columns=['Price']), df_test['Price']

In [12]:
# Preprocesar y filtrar los datos de prueba 
X_test_processed, y_test_processed = preprocessor.transform(X_test, y_test)
X_test_filtered, y_test_filtered = filter.transform(X_test_processed, y_test_processed)

X shape:  (8989, 40)
X_crossed_features shape:  (8989, 10)
X_EXPANDED shape:  (8989, 14)
Shape después de baja varianza en la transformación: (8989, 7)
Shape después de eliminar duplicados en la transformación: (8989, 7)
Shape después de eliminar correlacionadas en la transformación: (8989, 6)


In [ ]:
# Deshacer la transformación para obtener las predicciones originales 
y_test_pred = xgboost_model.predict(X_test_filtered)
y_test_pred_unscaled = preprocessor.inverse_transform(y_test_pred.reshape(-1, 1))

C:\Users\marti\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but QuantileTransformer was fitted with feature names
  warnings.warn(


In [14]:
from sklearn.metrics import ( root_mean_squared_error, 
                             mean_absolute_error, 
                             mean_absolute_percentage_error )

In [ ]:
y_true = y_test.values
y_pred = y_test_pred_unscaled
rmse = root_mean_squared_error(y_true, y_pred)
mae = mean_absolute_error(y_true, y_pred)
mape = mean_absolute_percentage_error(y_true, y_pred)


metrics = {
    "RMSE (₹)": f"{rmse:,.2f}",
    "MAE (₹)": f"{mae:,.2f}",
    "MAPE (%)": f"{mape:.2%}"
}


print("\nModel Performance Metrics:\n")
print(f"{'Metric':<15} {'Value':>15}")
print("-" * 30)
for metric, value in metrics.items():
    print(f"{metric:<15} {value:>15}")



Model Performance Metrics:

Metric                    Value
------------------------------
RMSE (₹)          25,817,976.00
MAE (₹)            6,926,706.50
MAPE (%)                 51.59%

Interpretation:
- RMSE: The model's predictions are typically off by ₹25,817,976.00 on average
- MAE: The average absolute error is ₹6,926,706.50
- MAPE: The predictions are off by 51.6% on average


Conclusiones sobre el rendimiento del nuevo modelo:
Evaluación de las métricas:
El nuevo modelo ha sido evaluado utilizando tres métricas clave de rendimiento: RMSE, MAE, y MAPE, y los resultados obtenidos son los siguientes:

RMSE (Root Mean Squared Error): 25,817,976.00

MAE (Mean Absolute Error): 6,926,706.50

MAPE (Mean Absolute Percentage Error): 51.59%

Estas métricas sugieren que el modelo tiene errores significativos, especialmente en términos absolutos (RMSE y MAE), aunque el valor de MAPE es considerablemente alto, lo que indica que el modelo se desvia un 51.59% de los valores reales en promedio.

Análisis Comparativo:
RMSE: El nuevo modelo tiene un RMSE más alto que el modelo anterior, lo que sugiere que las predicciones del nuevo modelo tienen un error absoluto mayor. 

MAE: Similar al RMSE, el MAE del nuevo modelo también es más alto, indicando que el nuevo modelo tiene un error promedio mayor en comparación con el modelo anterior.

MAPE: Aquí es donde se observa una gran diferencia. El MAPE de 51.59% en el nuevo modelo es mucho mayor que el 20.63% del modelo anterior. Esto sugiere que, en términos relativos, el nuevo modelo está cometiendo errores mucho mayores, lo cual es un indicativo de que el modelo tiene dificultades para generalizar correctamente en comparación con el modelo anterior.

Posibles causas de la diferencia en el rendimiento:
Sobreajuste (Overfitting): El nuevo modelo podría estar sobreajustado a los datos de entrenamiento, lo que significa que tiene un buen rendimiento en los datos de prueba, pero no generaliza bien a nuevos datos. Esto puede causar un mayor RMSE y MAE, así como un MAPE elevado.

Selección de características: El modelo anterior podría haber utilizado una mejor selección de características, mientras que el nuevo modelo podría estar incluyendo características irrelevantes o mal procesadas que afectan negativamente el rendimiento.

Hiperparámetros no optimizados: Los hiperparámetros del nuevo modelo podrían no estar correctamente ajustados, lo que podría estar limitando su capacidad para hacer buenas predicciones.